File: server/api/images.py
# my_app/api/images.py
from fastapi import APIRouter, Depends, Body, Query
from sqlalchemy.orm import Session
from typing import Optional, List
import os
import shutil

from server.db.database import get_db
from server.db.models import Image, Dataset, Class
from server.utils.string_utils import to_camel_case, to_snake_case

router = APIRouter()

@router.get("/")
def list_images(dataset_ids: Optional[List[str]] = Query(None), db: Session = Depends(get_db)):
    """
    - dataset_ids가 제공되면, 해당 데이터셋과 연결된 이미지 목록만 반환 (중복 제거)
    - dataset_ids가 없으면 전체 이미지를 반환
    - 각 이미지에는 연결된 클래스 목록(classIds)도 포함
    """
    if dataset_ids:
        # dataset_ids에 연결된 이미지들을 distinct()로 중복 제거
        images = db.query(Image).join(Image.datasets).filter(Dataset.id.in_(dataset_ids)).distinct().all()
    else:
        images = db.query(Image).all()
    
    results = []
    for img in images:
        results.append({
            "id": img.id,
            "filename": img.filename,
            "fileLocation": img.file_location,
            "thumbnailLocation": img.thumbnail_location,
            "width": img.width,
            "height": img.height,
            "approval": img.approval,
            "comment": img.comment,
            "classIds": [cls.id for cls in img.classes],
            "properties": img.properties,
        })
    
    return results

@router.get("/{image_id}")
def get_image(image_id: str, db: Session = Depends(get_db)):
    img = db.query(Image).filter(Image.id == image_id).first()
    if not img:
        return {"error": "Image not found"}
    return img

@router.post("/{image_id}")
def upsert_image(image_id: str, updated_data: dict = Body(...), db: Session = Depends(get_db)):
    # ✅ 요청 데이터(`camelCase` → `snake_case` 변환)
    updated_data = {to_snake_case(k): v for k, v in updated_data.items()}

    new_dataset_ids = updated_data.pop("dataset_ids", None)
    new_class_ids = updated_data.pop("class_ids", None)

    updated_data.setdefault("properties", {"description": "", "comment": ""})

    img = db.query(Image).filter(Image.id == image_id).first()

    if img:
        for field, value in updated_data.items():
            setattr(img, field, value)
        db.commit()
        db.refresh(img)
    else:
        img = Image(**updated_data)
        db.add(img)
        db.commit()
        db.refresh(img)

    if new_dataset_ids is not None:
        # 현재 연결된 dataset들의 ID를 가져오기
        existing_dataset_ids = {ds.id for ds in img.datasets}

        # 추가할 dataset 찾기 (이미 연결되지 않은 dataset만 추가)
        datasets_to_add = [db.query(Dataset).filter(Dataset.id == dataset_id).first()
                        for dataset_id in new_dataset_ids if dataset_id not in existing_dataset_ids]

        # dataset 추가
        img.datasets.extend(filter(None, datasets_to_add))  # None이 아닌 값만 추가

        db.commit()
        db.refresh(img)

    if new_class_ids is not None:
        existing_class_ids = {cls.id for cls in img.classes}

        # 추가할 class 찾기
        classes_to_add = [db.query(Class).filter(Class.id == class_id).first()
                        for class_id in new_class_ids if class_id not in existing_class_ids]

        # 기존 관계 제거 후 새 class만 추가
        img.classes.clear()
        img.classes.extend(filter(None, classes_to_add))

        db.commit()
        db.refresh(img)


    # ✅ 응답(`snake_case` → `camelCase` 변환)
    img_dict = img.__dict__.copy()
    response_data = {
        to_camel_case(k): v
        for k, v in img_dict.items()
        if not k.startswith("_")  # SQLAlchemy 내부 필드 제거
    }

    response_data["datasetIds"] = [ds.id for ds in img.datasets]
    response_data["classIds"] = [cls.id for cls in img.classes]

    return response_data

@router.delete("/{image_id}")
def delete_image(
    image_id: str,
    selected_dataset_ids: Optional[List[str]] = Query(None),  # 리스트로 받기
    db: Session = Depends(get_db)
):
    img = db.query(Image).filter(Image.id == image_id).first()
    if not img:
        return {"error": "Image not found"}
    
    print("selected_dataset_ids", selected_dataset_ids)

    # dataset_id가 제공된 경우
    if selected_dataset_ids:
        # image와 연결된 각 dataset에 대해
        for ds in list(img.datasets):
            if ds.id in selected_dataset_ids:
                ds.images.remove(img)  # 해당 dataset과의 연결만 제거
        db.commit()
        db.refresh(img)

        # 제거 후 이미지가 연결된 데이터셋이 없으면 파일 삭제 및 이미지 완전 삭제
        if len(img.datasets) == 0:
            _delete_image_files(img)
            db.delete(img)
            db.commit()
            return {"message": f"Image {image_id} fully deleted (no remaining dataset connections)."}
        else:
            return {"message": f"Image {image_id} unlinked from datasets {selected_dataset_ids}."}
    else:
        # dataset_id가 제공되지 않으면, 기본적으로 파일과 DB 모두에서 완전 삭제
        _delete_image_files(img)
        db.delete(img)
        db.commit()
        return {"message": f"Image {image_id} and associated files deleted."}

def _delete_image_files(img: Image):
    """이미지 파일과 썸네일을 삭제하는 유틸 함수"""
    if img.file_location and os.path.exists(img.file_location):
        try:
            os.remove(img.file_location)
        except Exception as e:
            print(f"Failed to delete file: {img.file_location}, Error: {e}")

    if img.thumbnail_location and os.path.exists(img.thumbnail_location):
        try:
            os.remove(img.thumbnail_location)
        except Exception as e:
            print(f"Failed to delete thumbnail: {img.thumbnail_location}, Error: {e}")

File: go_server/routes/images.go
package routes

import (
	"fmt"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"github.com/ingradient/ingradient/go_server/database"
	"github.com/ingradient/ingradient/go_server/dto"
	"github.com/ingradient/ingradient/go_server/models"
)

// ListImages (GET /api/images/)
func ListImages(c *gin.Context) {
	datasetIDs := c.QueryArray("dataset_ids") // ?dataset_ids=xxx&dataset_ids=yyy

	var images []models.Image
	var err error

	// Preload("Classes")를 사용하여 각 이미지의 연결된 Classes 데이터를 미리 로드함.
	if len(datasetIDs) > 0 {
		err = database.DB.Preload("Classes").
			Model(&models.Image{}).
			Joins("JOIN dataset_images di ON di.image_id = images.id").
			Where("di.dataset_id IN ?", datasetIDs).
			Distinct().
			Find(&images).Error
	} else {
		err = database.DB.Preload("Classes").Find(&images).Error
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// 필요한 데이터만 선택해서 반환. 예를 들어, 각 이미지에 연결된 클래스 ID 리스트(classIds)
	var results []map[string]interface{}
	for _, img := range images {
		// 연결된 클래스 목록에서 ID만 추출
		classIds := make([]string, 0, len(img.Classes))
		for _, cls := range img.Classes {
			classIds = append(classIds, cls.ID)
		}

		// 필요한 키들을 map에 담아서, 필요하다면 snake_case -> camelCase 변환(여기서는 수동 변환 예시)
		result := map[string]interface{}{
			"id":                img.ID,
			"filename":          img.Filename,
			"fileLocation":      img.FileLocation,
			"thumbnailLocation": img.ThumbnailLocation,
			"width":             img.Width,
			"height":            img.Height,
			"approval":          img.Approval,
			"comment":           img.Comment,
			"classIds":          classIds,
			"properties":        img.Properties,
			// "model":             img.ExtractedFeatures, // img.ExtractedFeatures가 정의되어 있다면
			"uploadAt":  img.UploadAt,
			"updatedAt": img.UpdatedAt,
		}
		results = append(results, result)
	}

	c.JSON(http.StatusOK, results)
}

// GetImage (GET /api/images/:image_id)
func GetImage(c *gin.Context) {
	imageID := c.Param("image_id")
	var img models.Image
	if err := database.DB.Where("id = ?", imageID).First(&img).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Image not found"})
		return
	}

	c.JSON(http.StatusOK, img)
}

// UpsertImage (POST /api/images/:image_id)
func UpsertImage(c *gin.Context) {
	// URL 파라미터로 이미지 ID 받기
	imageID := c.Param("image_id")

	// 1) JSON 바디 → ImageUpsertReq 구조체
	var req dto.ImageUpsertReq
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// 2) DB에서 이미지 조회
	var img models.Image
	if err := database.DB.Where("id = ?", imageID).First(&img).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			// (2-1) 레코드 없으면 새로 생성
			img = models.Image{
				ID:                imageID,
				Filename:          req.Filename,
				FileLocation:      req.FileLocation,
				ThumbnailLocation: req.ThumbnailLocation,
				Approval:          req.Approval,
				Comment:           req.Comment,
				Height:            req.Height,
				Width:             req.Width,
				Type:              req.Type,
				Size:              req.Size,
			}
			if errCreate := database.DB.Create(&img).Error; errCreate != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": errCreate.Error()})
				return
			}
		} else {
			// (2-2) 다른 DB 에러
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
	} else {
		// (2-3) 기존 레코드 업데이트
		img.Filename = req.Filename
		img.FileLocation = req.FileLocation
		img.ThumbnailLocation = req.ThumbnailLocation
		img.Approval = req.Approval
		img.Comment = req.Comment
		img.Height = req.Height
		img.Width = req.Width
		img.Type = req.Type
		img.Size = req.Size

		if errSave := database.DB.Save(&img).Error; errSave != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": errSave.Error()})
			return
		}
	}

	// 3) Many2Many 관계: Datasets
	if req.DatasetIDs != nil {
		database.DB.Model(&img).Association("Datasets").Clear()
		for _, dsID := range req.DatasetIDs {
			var ds models.Dataset
			if err := database.DB.Where("id = ?", dsID).First(&ds).Error; err == nil {
				database.DB.Model(&img).Association("Datasets").Append(&ds)
			}
		}
	}

	// 4) Many2Many 관계: Classes
	if req.ClassIDs != nil {
		database.DB.Model(&img).Association("Classes").Clear()
		for _, clID := range req.ClassIDs {
			var classObj models.Class
			if err := database.DB.Where("id = ?", clID).First(&classObj).Error; err == nil {
				database.DB.Model(&img).Association("Classes").Append(&classObj)
			}
		}
	}

	// 5) 최종 응답
	c.JSON(http.StatusOK, img)
}

// DeleteImage (DELETE /api/images/:image_id)
func DeleteImage(c *gin.Context) {
	imageID := c.Param("image_id")
	selectedDatasetIDs := c.QueryArray("selected_dataset_ids") // ?selected_dataset_ids=xxx

	var img models.Image
	if err := database.DB.Where("id = ?", imageID).First(&img).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Image not found"})
		return
	}

	if len(selectedDatasetIDs) > 0 {
		// 특정 dataset과의 연결만 해제
		// 1) img.datasets 중에서 selectedDatasetIDs에 해당하는 부분만 Remove
		for _, dsID := range selectedDatasetIDs {
			var ds models.Dataset
			if err := database.DB.Where("id = ?", dsID).First(&ds).Error; err == nil {
				database.DB.Model(&img).Association("Datasets").Delete(&ds)
			}
		}
		// 연결 해제 후 여전히 연결된 dataset이 없으면 이미지 삭제
		// 다시 DB에서 refresh
		database.DB.Where("id = ?", imageID).First(&img)
		// Preload datasets
		dsCount := database.DB.Model(&img).Association("Datasets").Count()
		if dsCount == 0 {
			// 완전 삭제
			deleteImageFiles(img)
			database.DB.Delete(&img)
			c.JSON(http.StatusOK, gin.H{"message": fmt.Sprintf("Image %s fully deleted", imageID)})
			return
		}
		c.JSON(http.StatusOK, gin.H{"message": fmt.Sprintf("Image %s unlinked from datasets %v", imageID, selectedDatasetIDs)})
	} else {
		// dataset_id 지정이 없으면 파일+DB 완전 삭제
		deleteImageFiles(img)
		database.DB.Delete(&img)
		c.JSON(http.StatusOK, gin.H{"message": fmt.Sprintf("Image %s and associated files deleted", imageID)})
	}
}

func deleteImageFiles(img models.Image) {
	// 파일 삭제
	if img.FileLocation != "" {
		if err := os.Remove(img.FileLocation); err != nil {
			fmt.Println("Failed to delete file:", err)
		}
	}
	if img.ThumbnailLocation != "" {
		if err := os.Remove(img.ThumbnailLocation); err != nil {
			fmt.Println("Failed to delete thumbnail:", err)
		}
	}
}
